### PR TITLE
Remove PROCESS GROUP rpc backend and process group agent code

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61958 Remove PROCESS GROUP rpc backend and process group agent code**
* **#61907 Remove faulty process group code**
* **#61421 Add faulty tensorpipe implementation**

Remove process_group_agent.cpp and relevant process group agent code. The BackendType PROCESS_GROUP will still be available to the user, but will throw an error if they try to use it.

Differential Revision: [D29819343](https://our.internmc.facebook.com/intern/diff/D29819343)

### CODE REVIEW

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29819343).

File: torch/distributed/rpc/__init__.py [hunk 1]
```diff
@@ -159,13 +156,12 @@ def init_rpc(
             backend = BackendType.TENSORPIPE  # type: ignore[attr-defined]
 
         if backend == BackendType.PROCESS_GROUP:  # type: ignore[attr-defined]
```
Comment (mrshenli) [inline torch/distributed/rpc/__init__.py:158]: Do we also plan to remove `PROCESS_GROUP` from `BackendType`?

Comment (H-Huang) [reply torch/distributed/rpc/__init__.py:158]: We can remove it and the user will get an `AttributeError` if they have existing code that uses it. Or we can keep the enum type (remove rest of code) for a release and throw a runtime error with description when the user tries to use it. Is it better to just remove it altogether?

Comment (mrshenli) [reply torch/distributed/rpc/__init__.py:158]: Make sense. Keeping the enum item for one more release might be more user-friendly.

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29819343).

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29819343).

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29819343).

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29819343).

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29819343).

**Comment by H-Huang** [comment]
closing and will recreate due to revert of a previous PR

**Pull request closed without merging**